component/kafka/kafka.go

```diff
@@ -0,0 +1,104 @@
+// Package kafka provides some shared interfaces for the Kafka components.
+package kafka
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+
+	"github.com/Shopify/sarama"
+	"github.com/opentracing/opentracing-go"
+)
+
+// FailStrategy type definition.
+type FailStrategy int
+
+const (
+	// ExitStrategy does not commit failed message offsets and exits the application.
+	ExitStrategy FailStrategy = iota
+	// SkipStrategy commits the offset of messages that failed processing, and continues processing.
+	SkipStrategy
+)
+
+// BatchProcessorFunc definition of a batch async processor function.
+type BatchProcessorFunc func(Batch) error
+
+// Message interface for wrapping messages that are handled by the kafka component.
+type Message interface {
+	// Context will contain the context to be used for processing.
+	// Each context will have a logger setup which can be used to create a logger from context.
+	Context() context.Context
+	// Message will contain the raw Kafka message.
+	Message() *sarama.ConsumerMessage
+	// Span contains the tracing span of this message.
+	Span() opentracing.Span
+}
+
+// NewMessage initializes a new message which is an implementation of the kafka Message interface
+func NewMessage(ctx context.Context, sp opentracing.Span, msg *sarama.ConsumerMessage) Message {
+	return &message{
+		ctx: ctx,
```

Comment:
Do you think we could get rid of this? I just noticed it goes straight into the face of https://blog.golang.org/context-and-structs

Reply:
Initially we discussed to approach a per message commit logic as we had it in SQS and AMQP.
Now we changed the approach so we might be able to remove this?

Reply:
And better change the `type BatchProcessorFunc func(Batch) error` to `type BatchProcessorFunc func(context.Context, Batch) error`?

Reply:
But with patron that can't be because every message has it's correlation ID 
```
ctx, sp := c.getContextWithCorrelation(msg)
```

We will then disable features like `log.WithContext(...)` for every message in the client.

Reply:
Under **Storing context in structs leads to confusion**

> The (*Worker).Fetch and (*Worker).Process method both use a context stored in Worker. This prevents the callers of Fetch and Process (which may themselves have different contexts) from specifying a deadline, requesting cancellation, and attaching metadata on a per-call basis. For example: the user is unable to provide a deadline just for (*Worker).Fetch, or cancel just the (*Worker).Process call. The caller's lifetime is intermingled with a shared context, and the context is scoped to the lifetime where the Worker is created.

For the message, I don't think we have this problem at all. The context is there and it's clear how it will be used.

Reply:
Let's leave it then.